### ISSUE TITLE
Better wording on dynamic blocks with sensitive value lists

### ISSUE THREAD

**Issue by wonko** [opentofu/opentofu#1434]
### OpenTofu Version

```shell
➜ tofu version
OpenTofu v1.6.1
on darwin_arm64
```

### OpenTofu Configuration Files

```hcl
resource "kubernetes_role_binding" "foo_service_access_binding" {
  metadata {
    name      = "service-access-binding"
    namespace = "foo"
  }

  role_ref {
    api_group = "rbac.authorization.k8s.io"
    kind      = "Role"
    name      = "bar"
  }

  dynamic "subject" {
    for_each = var.developer_subjects

    content {
      kind      = "User"
      name      = subject.value
      api_group = "rbac.authorization.k8s.io"
    }
  }
}
```

With `var.developer_subjects` being an array of aws_user.arn fields, which are sensitive (through the use of the aws_iam module).

### Debug Output

none needed, it's the wording of the error, not running tofu itself.

### Expected Behavior

The error should have been:

```
│ Error: Invalid dynamic for_each value
│ 
│   on ../../modules/.../rbac.tf line 122, in resource "kubernetes_role_binding" "foo_status_access_binding":
│  122:     for_each = var.developer_subjects
│ 
│ Cannot use a sensitive value in for_each.
```

### Actual Behavior

Plan/apply doesn't work, and gives the following error:

```
│ Error: Invalid dynamic for_each value
│ 
│   on ../../modules/.../rbac.tf line 122, in resource "kubernetes_role_binding" "foo_status_access_binding":
│  122:     for_each = var.developer_subjects
│ 
│ Cannot use a list of string value in for_each. An iterable collection is required.
```

### Steps to Reproduce

Run any dynamic block on a sensitive list.

### Additional Context

Solved it by using a `nonsensitive` argument on the list. The functionality then works. I've lost the best part of an hour trying to get my head around this...

### References

https://github.com/hashicorp/terraform/issues/32828 is the issue at TF, and https://github.com/hashicorp/terraform/pull/34706 seems a stab at a solution.

**Comment by Evi1Pumpkin** [comment]
Hi @wonko. Thank you for opening this issue. 
That's indeed a confusing output.

Our team regularly reviews new issues, but this process may take some time. We appreciate your patience as we work through your issue. If you're interested, the contribution guide includes a section on our [decision-making process](https://github.com/opentofu/opentofu/blob/main/CONTRIBUTING.md#who-decides-if-a-feature-will-be-implemented-and-how-is-that-decision-made).

**Comment by chenzejun** [comment]
Hi @Evi1Pumpkin @wonko, I have a background in it.

During the `evaluateForEachExpressionValue` we check two things:

1. If the whole foe_each value is sensitive
2. If the foe_each value is suitable type

And 1 will happen before 2, so the diagnostic message of `developer_subjects` in this issue only tells it is sensitive.


We have two ways to fix this issue：
1. Do 2 before 1, then we get:
```
│ Error: Invalid dynamic for_each value
│ 
│   on ../../modules/.../rbac.tf line 122, in resource "kubernetes_role_binding" "foo_status_access_binding":
│  122:     for_each = var.developer_subjects
│ 
│ Sensitive values, or values derived from sensitive values, cannot be used as for_each arguments. If used, the sensitive value could be exposed as a resource instance key.

```
  
2. Do both, then we get:
```
│ Error: Invalid dynamic for_each value
│ 
│   on ../../modules/.../rbac.tf line 122, in resource "kubernetes_role_binding" "foo_status_access_binding":
│  122:     for_each = var.developer_subjects
│ 
│ Sensitive values, or values derived from sensitive values, cannot be used as for_each arguments. If used, the sensitive value could be exposed as a resource instance key.


│ Error: Invalid dynamic for_each value
│ 
│   on ../../modules/.../rbac.tf line 122, in resource "kubernetes_role_binding" "foo_status_access_binding":
│  122:     for_each = var.developer_subjects
│ 
│ The given "for_each" argument value is unsuitable: the "for_each" argument must be a map, or set of strings, and you have provided a value of type tuple.
```

I prefer 1, and I'am happy to help fix it.

**Comment by wonko** [comment]
In option 1, the error message `The given "for_each" argument value is unsuitable: the "for_each" argument must be a map, or set of strings, and you have provided a value of type tuple.` still doesn't address the fact that the argument is sensitive.

The first message given in option two is actually what is the actionable message, where it is immediately clear what is "wrong".

So I'd prefer option two - even with the two messages.

**Comment by cam72cam** [comment]
I'm also a fan of 2 as well.  It conveys both pieces of information, which is the advantage of the diagnostics system.  I'll let @Evi1Pumpkin way in tomorrow.  If she agrees, I think we can accept this issue and assign to @chenzejun if you'd like.

**Comment by Evi1Pumpkin** [comment]
I agree with everything you said and also prefer the more explicit number 2.

**Comment by cam72cam** [comment]
@chenzejun thanks for taking this on, assigned!

**Comment by chenzejun** [comment]
@cam72cam @Evi1Pumpkin Hi guys! PTAL https://github.com/opentofu/opentofu/pull/1485
